# Analysis exploratorio - Titanic

Vamos a realizar un analisis para poder entender nuestros datos

## 1 Cargamos y describimos la data

In [1]:
# Cargamos las librerias necesarias
import numpy as np
import pandas as pd

In [2]:
data = pd.read_csv("cienciadedatosenaws/episodio1/data.csv")

In [3]:
data

PassengerId  Survived  Pclass  \
0            892         0       3   
1            893         1       3   
2            894         0       2   
3            895         0       3   
4            896         1       3   
..           ...       ...     ...   
413         1305         0       3   
414         1306         1       1   
415         1307         0       3   
416         1308         0       3   
417         1309         0       3   

                                             Name     Sex   Age  SibSp  Parch  \
0                                Kelly, Mr. James    male  34.5      0      0   
1                Wilkes, Mrs. James (Ellen Needs)  female  47.0      1      0   
2                       Myles, Mr. Thomas Francis    male  62.0      0      0   
3                                Wirz, Mr. Albert    male  27.0      0      0   
4    Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female  22.0      1      1   
..                                            ...     ...   ...    ...    ...   
413                            Spector, Mr. Woolf    male   NaN      0      0   
414                  Oliva y Ocana, Dona. Fermina  female  39.0      0      0   
415                  Saether, Mr. Simon Sivertsen    male  38.5      0      0   
416                           Ware, Mr. Frederick    male   NaN      0      0   
417                      Peter, Master. Michael J    male   NaN      1      1   

                 Ticket      Fare Cabin Embarked  
0                330911    7.8292   NaN        Q  
1                363272    7.0000   NaN        S  
2                240276    9.6875   NaN        Q  
3                315154    8.6625   NaN        S  
4               3101298   12.2875   NaN        S  
..                  ...       ...   ...      ...  
413           A.5. 3236    8.0500   NaN        S  
414            PC 17758  108.9000  C105        C  
415  SOTON/O.Q. 3101262    7.2500   NaN        S  
416              359309    8.0500   NaN        S  
417                2668   22.3583   NaN        C  

[418 rows x 12 columns]

In [4]:
data.shape

(418, 12)

In [5]:
data.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [6]:
data.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   418.000000  418.000000  418.000000  332.000000  418.000000   
mean   1100.500000    0.363636    2.265550   30.272590    0.447368   
std     120.810458    0.481622    0.841838   14.181209    0.896760   
min     892.000000    0.000000    1.000000    0.170000    0.000000   
25%     996.250000    0.000000    1.000000   21.000000    0.000000   
50%    1100.500000    0.000000    3.000000   27.000000    0.000000   
75%    1204.750000    1.000000    3.000000   39.000000    1.000000   
max    1309.000000    1.000000    3.000000   76.000000    8.000000   

            Parch        Fare  
count  418.000000  417.000000  
mean     0.392344   35.627188  
std      0.981429   55.907576  
min      0.000000    0.000000  
25%      0.000000    7.895800  
50%      0.000000   14.454200  
75%      0.000000   31.500000  
max      9.000000  512.329200

Vamos a ver que variables categoricas tenemos en el dataset

In [7]:
categorical = data.dtypes[data.dtypes == "object"].index
print(categorical)
data[categorical].describe()

Index(['Name', 'Sex', 'Ticket', 'Cabin', 'Embarked'], dtype='object')


Name   Sex    Ticket            Cabin Embarked
count                418   418       418               91      418
unique               418     2       363               76        3
top     Kelly, Mr. James  male  PC 17608  B57 B59 B63 B66        S
freq                   1   266         5                3      270

In [8]:
data[['Name', 'Sex', 'Ticket', 'Cabin', 'Embarked']].describe()

Name   Sex    Ticket            Cabin Embarked
count                418   418       418               91      418
unique               418     2       363               76        3
top     Kelly, Mr. James  male  PC 17608  B57 B59 B63 B66        S
freq                   1   266         5                3      270

## 2 ¿Necesitamos todas las variables?

PassengerId: Es un numero asignado de manera aleatoria, no es muy probable que nos vaya a ofrecer informacion relevante

Survived: Indica si un pasajero sobrevivio o no

Las variables que describen mejor numericamente a los pasajeros son: Pclass, Sex, Age, SibSp, Parch, Fare y Embarked

Tenemos 3 categorias para considerar: Name, Ticket and Cabin.

In [9]:
data["Name"].head(15)

0                                      Kelly, Mr. James
1                      Wilkes, Mrs. James (Ellen Needs)
2                             Myles, Mr. Thomas Francis
3                                      Wirz, Mr. Albert
4          Hirvonen, Mrs. Alexander (Helga E Lindqvist)
5                            Svensson, Mr. Johan Cervin
6                                  Connolly, Miss. Kate
7                          Caldwell, Mr. Albert Francis
8             Abrahim, Mrs. Joseph (Sophie Halaut Easu)
9                               Davies, Mr. John Samuel
10                                     Ilieff, Mr. Ylio
11                           Jones, Mr. Charles Cresson
12        Snyder, Mrs. John Pillsbury (Nelle Stevenson)
13                                 Howard, Mr. Benjamin
14    Chaffee, Mrs. Herbert Fuller (Carrie Constance...
Name: Name, dtype: object

In [10]:
data["Name"].describe()

count                  418
unique                 418
top       Kelly, Mr. James
freq                     1
Name: Name, dtype: object

In [11]:
data["Ticket"].head(15) 

0          330911
1          363272
2          240276
3          315154
4         3101298
5            7538
6          330972
7          248738
8            2657
9       A/4 48871
10         349220
11            694
12          21228
13          24065
14    W.E.P. 5734
Name: Ticket, dtype: object

In [12]:
data["Ticket"].describe()

count          418
unique         363
top       PC 17608
freq             5
Name: Ticket, dtype: object

In [13]:
data["Cabin"].head(15)

0     NaN
1     NaN
2     NaN
3     NaN
4     NaN
5     NaN
6     NaN
7     NaN
8     NaN
9     NaN
10    NaN
11    NaN
12    B45
13    NaN
14    E31
Name: Cabin, dtype: object

In [14]:
data["Cabin"].describe()

count                  91
unique                 76
top       B57 B59 B63 B66
freq                    3
Name: Cabin, dtype: object

Eliminamos las columnas que no ayudan a nuestro analisis

In [15]:
del data["PassengerId"]
del data["Ticket"]

## 3 ¿Deberiamos transformar algunas variables?

Survived: deberiamos convertirlo a una variable categorica

In [16]:
new_survived = pd.Categorical(data["Survived"])
print(new_survived)
new_survived = new_survived.rename_categories(["Died","Survived"])
new_survived

[0, 1, 0, 0, 1, ..., 0, 1, 0, 0, 0]
Length: 418
Categories (2, int64): [0, 1]


['Died', 'Survived', 'Died', 'Died', 'Survived', ..., 'Died', 'Survived', 'Died', 'Died', 'Died']
Length: 418
Categories (2, object): ['Died', 'Survived']

In [17]:
new_survived.describe()

counts     freqs
categories                  
Died           266  0.636364
Survived       152  0.363636

In [18]:
data["Survived"] = new_survived

Pclass: Indica la clase del pasajero

In [19]:
new_Pclass = pd.Categorical(data["Pclass"],
                           ordered=True)
new_Pclass

[3, 3, 2, 3, 3, ..., 3, 1, 3, 3, 3]
Length: 418
Categories (3, int64): [1 < 2 < 3]

In [20]:
new_Pclass = new_Pclass.rename_categories(["Class1","Class2","Class3"])    
new_Pclass

['Class3', 'Class3', 'Class2', 'Class3', 'Class3', ..., 'Class3', 'Class1', 'Class3', 'Class3', 'Class3']
Length: 418
Categories (3, object): ['Class1' < 'Class2' < 'Class3']

In [21]:
new_Pclass.describe()

counts     freqs
categories                  
Class1         107  0.255981
Class2          93  0.222488
Class3         218  0.521531

In [22]:
data["Pclass"] = new_Pclass

Cabin: Indica la cabian del pasajero

In [ ]:
data["Cabin"].unique() 

In [ ]:
char_cabin = data["Cabin"].astype(str) 
new_Cabin = np.array([cabin[0] for cabin in char_cabin])
new_Cabin = pd.Categorical(new_Cabin)

In [ ]:
new_Cabin .describe()

In [ ]:
data["Cabin"] = new_Cabin

## 4 ¿Existen valores de NA, valores atípicos u otros valores extraños?

In [ ]:
data.isnull().sum()

In [ ]:
data["Age"].describe()

Observe que el recuento de edad (712) es menor que el recuento total de filas del conjunto de datos (889). Esto indica datos faltantes. Podemos obtener los índices de fila de los valores faltantes con np.where()

In [ ]:
missing = np.where(data["Age"].isnull() == True)
missing

In [ ]:
len(missing[0])

Hay 86 valores nulos, no es una buena idea eliminar la columna, por que la edad es un factor importante para el analisis de este conjunto de datos. Podriamos tomar varias desiciones para llenar estos datos, algunos pueden ser:

Imputacion de datos
- Llenar los valores con 0s
- Llenar los valores con la media
- Llenar los valores con la mediana

Para escojer la mejor solucion vamos a ver el histograma de la edad

In [ ]:
data.hist(column='Age',             # Columna que deseamos graficar
                   figsize=(9,6),   # Tamaño de la figura
                   bins=20)         # Numero de cajas

La mayoria de edades aproximadamente se concentran en 22 y 26, asi que elegiremos el numero 24 para llenar los valores nulos 

In [ ]:
new_age_var = np.where(data["Age"].isnull(), # Logical check
                       28,                       # Value if check is true
                       data["Age"])     # Value if check is false

data["Age"] = new_age_var 

In [ ]:
data["Age"].describe()

## 5 ¿Debeiamos crear nuevas variables?

Creemos una nueva variable, Familia, que combine SibSp y Parch para indicar el número total de familiares (hermanos, cónyuges, padres e hijos) que tiene un pasajero a bordo

In [ ]:
data["Family"] = data["SibSp"] + data["Parch"]
most_family = np.where(data["Family"] == max(data["Family"]))
most_family

In [ ]:
data.loc[most_family]

## 6 Conclusiones

- La mayoria de pasajero tiene entre 30 y 20 años
- 152 Sobrevivientes y 266 fallecidos